src/command/list/toolchain.rs

```diff
@@ -0,0 +1,345 @@
+        inventory: &Inventory,
+        project: &Option<Rc<Project>>,
+        filter: &Filter,
+    ) -> Fallible<Toolchain> {
+        /// An internal-only helper for tracking whether we found a given item
+        /// from the `PackageCollection` as a *package* or as a *tool*.
+        #[derive(PartialEq, Debug)]
+        enum Kind {
+            Package,
+            Tool,
+        }
+
+        /// A convenient name for this tuple, since we have to name it in a few
+        /// spots below.
+        type Triple = (Kind, PackageConfig, Source);
+
+        let (packages_and_tools, errors): (Vec<Fallible<Triple>>, Vec<Fallible<Triple>>) =
+            inventory
+                .packages
+                .clone()
+                .into_iter()
+                .filter_map(|config| {
+                    // Start with the package itself, since tools often match
+                    // the package name and we prioritize packages.
+                    if &config.name == name {
+                        let source = Package::source(name, &config.version, project);
+                        if source.allowed_with(filter) {
+                            Some(Ok((Kind::Package, config, source)))
+                        } else {
+                            None
+                        }
+
+                    // Then check if the passed name matches an installed package's
+                    // binaries. If it does, we have a tool.
+                    } else if config
+                        .bins
+                        .iter()
+                        .find(|bin| bin.as_str() == name)
+                        .is_some()
+                    {
+                        tool_source(name, &config.version, project)
+                            .map(|source| {
+                                if source.allowed_with(filter) {
+                                    Some((Kind::Tool, config, source))
+                                } else {
+                                    None
+                                }
+                            })
+                            .transpose()
+
+                    // Otherwise, we don't have any match all.
+                    } else {
+                        None
+                    }
+                })
+                .partition(|result| result.is_ok());
+
+        // If there are any errors, return the first one, since we don't have a
+        // notion of composite errors in our error handling at this point.
+        // (This should be quite unusual, as it means something failed in
+        // looking up a tool's bin directory.)
+        let first_error = errors.into_iter().map(|error| error.unwrap_err()).nth(0);
+        if let Some(error) = first_error {
+            return Err(error);
+        }
+
+        let packages_and_tools: Vec<Triple> = packages_and_tools
+            .into_iter()
+            .map(|ok| ok.unwrap())
+            .collect();
+
+        let has_packages = packages_and_tools
```

Comment:
When fixing the above, I also realized that I could simplify this; it cleans up really nicely if I use `fold` and a `(bool, bool)` instead of iterating over the `Vec` twice.

Reply:
:+1: